File: android/controller/src/main/java/org/openbot/controller/ForwardSpeed.kt
package org.openbot.controller

import android.annotation.SuppressLint
import android.util.Log
import androidx.annotation.FloatRange
import kotlin.math.roundToInt

object ForwardSpeed {
    const val max = 1f
    const val min = 0f
    const val minNegative = -1f

    @FloatRange(from=min.toDouble(),to=max.toDouble())
    var value = 0f

    fun increment(incrementValue: Float) {
        value = max.coerceAtMost(value + incrementValue).round(2)
    }

    fun decrement(decrementValue: Float) {
        value = min.coerceAtLeast(value - decrementValue).round(2)
    }

    fun decrementNegative(decrementValue: Float) {
        value = minNegative.coerceAtLeast(value - decrementValue).round(2)
    }

    fun reset() {
        value = 0f
    }

    fun setTo(minSpeed: Float) {
        if (minSpeed !in -1f..1f) {
            Log.e("ForwardSpeed", "setTo () got invalid parameter $minSpeed")
            return
        }
        value = minSpeed
    }

    fun isMax(): Boolean {
        return value >= max
    }

    fun isMin(): Boolean {
        return value <= min
    }

    private fun Float.round(decimals: Int): Float {
        var multiplier:Int = 1
        repeat(decimals) { multiplier *= 10 }
        return kotlin.math.round(this * multiplier) / multiplier
    }
}

File: controller/flutter/lib/utils/forwardSpeed.dart
import 'dart:math';

class ForwardSpeed {
  static const max = 1.0;
  static const min = 0.0;
  static const minNegative = -1.0;
  static double value = 0;

  static void increment(double incrementValue) {
    if (value + incrementValue > max) {
      value = max;
    } else {
      value = (value + incrementValue).toPrecision(2);
    }
  }

  static void decrement(double decrementValue) {
    if (value - decrementValue < min) {
      value = (value - decrementValue).toPrecision(2);
    } else {
      value = min;
    }
  }

  static void decrementNegative(double decrementValue) {
    if (value + decrementValue > minNegative) {
      value = (value + decrementValue).toPrecision(2);
    } else {
      value = minNegative;
    }
  }

  static void reset() {
    value = 0;
  }

  static void setTo(double minSpeed) {
    if (minSpeed < min && minSpeed > max) {
      return;
    }
    value = minSpeed;
  }

  static bool isMin() {
    return value <= min;
  }
}

extension Precision on double {
  double toPrecision(int fractionDigits) {
    num mod = pow(10, fractionDigits.toDouble());
    return ((this * mod).round().toDouble() / mod);
  }
}
